versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/AbstractConcurrency.java

```diff
@@ -0,0 +1,235 @@
+                    commitAttempt.putGlobal(
+                        contentsId,
+                        ByteString.copyFromUtf8(
+                            Integer.toString(
+                                Integer.parseInt(currentStates.get(ki).toStringUtf8()) + 1)));
+                    if (!variation.sharedKeys) {
+                      commitAttempt.putExpectedStates(
+                          contentsId, Optional.of(currentStates.get(ki)));
+                    }
+                    commitAttempt.addPuts(
+                        KeyWithBytes.of(keys.get(ki), contentsId, (byte) 0, ByteString.EMPTY));
+                  }
+
+                  try {
+                    commitAttempt
+                        .commitToBranch(branch)
+                        .commitMetaSerialized(
+                            ByteString.copyFromUtf8(
+                                "commit #"
+                                    + commit
+                                    + " to "
+                                    + branch.getName()
+                                    + " something "
+                                    + ThreadLocalRandom.current().nextLong()));
+                    databaseAdapter.commit(commitAttempt.build());
+                    commitsOK.incrementAndGet();
+                  } catch (ReferenceRetryFailureException retry) {
+                    retryFailures.incrementAndGet();
+                  }
+                }
+              } catch (Exception e) {
+                e.printStackTrace();
+                throw new RuntimeException(e);
+              }
+            });
+      }
+
+      for (Entry<BranchName, Set<Key>> branchKeys : keysPerBranch.entrySet()) {
+        BranchName branch = branchKeys.getKey();
+        databaseAdapter.create(branch, databaseAdapter.toHash(BranchName.of("main")));
+        ImmutableCommitAttempt.Builder commitAttempt =
+            ImmutableCommitAttempt.builder()
+                .commitToBranch(branchKeys.getKey())
+                .commitMetaSerialized(
+                    ByteString.copyFromUtf8("initial commit for " + branch.getName()));
+        for (Key k : branchKeys.getValue()) {
+          ContentsId contentsId = keyToContentsId.get(k);
+          commitAttempt.addPuts(KeyWithBytes.of(k, contentsId, (byte) 0, ByteString.EMPTY));
+          commitAttempt.putGlobal(contentsId, ByteString.copyFromUtf8("0"));
+        }
+        databaseAdapter.commit(commitAttempt.build());
+      }
+
+      tasks.forEach(executor::submit);
+
+      startLatch.countDown();
+      Thread.sleep(1_500);
+      stopFlag.set(true);
+
+      // 30 seconds is long, but necessary to let transactional databases detect deadlocks, which
+      // cause Nessie-commit-retries.
+      assertThat(stopLatch.await(30, TimeUnit.SECONDS)).isTrue();
+
+    } finally {
+      stopFlag.set(true);
+
+      System.out.printf(
+          "AbstractTieredCommitsTest.concurrency - %s : Commits OK: %s  Retry-Failures: %s%n",
```

Comment:
do we expect a certain % of failures? Is it worth printing out results of a unit test if we can't do anything about it?

Reply:
It's tricky to predict a certain percentage here.
While transactional DBs (SQL) do have a very low failure rate, it is still > 0 though. And that rate also depends on the performance of the system.  And of course depends on the test configuration (shared vs distinct branches / keys).